File: nngpp/reqrep_time/responder/responder.cc
// Copyright (C) 2023 twyleg
#include <utils/time.h>

#include <nngpp/nngpp.h>
#include <nngpp/protocol/rep0.h>
#include <nngpp/platform/platform.h>

#include <rapidjson/rapidjson.h>
#include <rapidjson/writer.h>

#include <string>
#include <chrono>
#include <thread>
#include <iostream>
#include <iomanip>


const std::string address = "ipc:///tmp/reqrep_time_date.ipc";


int main(int argc, char *argv[]) {
	std::cout << "Started reqrep_time_responder\n\r";


	auto res = nng::rep::open();
	res.listen(address.c_str());

	int i = 0;
	while(true) {

		auto msg = res.recv_msg();

		const nng::view msg_view = msg.body().get();
		const auto request = std::basic_string_view(msg_view.data<char>(), msg_view.size());

		std::cout << "Received request: \"" << request << "\"" << std::endl;

		if(request == "time") {
			const auto time = getJsonTimeString();
			res.send(nng::view(time.data(), time.size()));
			std::cout << "Responded with time: " << time << std::endl;

		} else if(request == "date") {
			const auto date = getJsonDateString();
			res.send(nng::view(date.data(), date.size()));
			std::cout << "Responded with date: " << date << std::endl;
		}
	}

	return 0;
}

File: pynng/reqrep_time/responder.py
# Copyright (C) 2023 twyleg
import json
from datetime import datetime

import pynng


address = "ipc:///tmp/reqrep_time_date.ipc"


def get_json_time_string_as_bytes() -> bytes:
    current_time = datetime.now()

    time_date = {
        "time": current_time.strftime("%X"),
    }

    return json.dumps(time_date).encode()


def get_json_date_string_as_bytes() -> bytes:
    current_time = datetime.now()

    time_date = {
        "date": current_time.strftime("%x"),
    }

    return json.dumps(time_date).encode()


if __name__ == '__main__':

    with pynng.Rep0(listen=address) as rep:

        while True:

            request_msg = rep.recv_msg()
            request = request_msg.bytes.decode()

            print(f"Received request: {request}")
            match request:
                case "time":
                    time = get_json_time_string_as_bytes()
                    rep.send(time)
                    print(f"Responded with time: {time.decode()}")
                case "date":
                    date = get_json_date_string_as_bytes()
                    rep.send(date)
                    print(f"Responded with date: {date.decode()}")
